examples/ReAct/hotpotqa/inference.py

```diff
@@ -0,0 +1,184 @@
+
+        state = copy.deepcopy(state)
+        current_state = state.current_state
+        step_idx = state.step_idx
+        try:
+            thought, action = action.strip().split(f"\nAction {step_idx}: ")
+            if "Search" in action:
+                new_state = current_state + self.toolset[0](env=self.base_model, step_idx=step_idx, action=action, thought=thought)
+                print("Search tool is used")
+            elif "Lookup" in action:
+                new_state = current_state + self.toolset[1](env=self.base_model, step_idx=step_idx, action=action, thought=thought)
+                print("Lookup tool is used")
+            elif "Finish" in action:
+                new_state = current_state + self.toolset[2](env=self.base_model, step_idx=step_idx, action=action, thought=thought)
+                print("Finish tool is used")
+            else:
+                self.terminal = True
+                new_state =  None
+        except:
+            thought = action.strip().split('\n')[0]
+            if step_idx != 1 and thought == "":
+                self.terminal = True
+                return None
+            new_state = current_state + f" {thought}\nAction {step_idx}:"
+
+        current_state = new_state
+        print(thought,action)
+        action = action
+
+        if action == "exceed maxlength":
+            self.terminal = True
+
+        if self.terminal == True:
+            state = HotpotqaState(step_idx=self.max_steps, last_state=None,
+                        current_state=None, action= "Finish[]")
+            aux = {"ans_finished": self.terminal}
+            return state, aux
+
+        state = HotpotqaState(step_idx=step_idx+1, last_state=state.current_state,
+                        current_state=current_state, action=action)
+        aux = {"ans_finished": self.terminal}
+
+        return state, aux
+
+    def is_terminal(self, state: HotpotqaState) -> bool:
+        if self.terminal == True:
+            self.base_model.reset()
+            self.terminal = False
+            return True
+        if utils.finished_check(state.action):
+            self.base_model.reset()
+            self.terminal = False
+            return True
+        if state.step_idx == self.max_steps:
+            self.base_model.reset()
+            self.terminal = False
+            return True
+        return False
+
+def main(model_dir, llama_size="8B", prompt="examples/ReAct/hotpotqa/prompts/react.json", data_path="examples/ReAct/hotpotqa/data/hotpot_dev_v1_simplified.json"):
+    
+    base_model = Llama3Model(model_dir, llama_size, max_batch_size=1, max_seq_len=20000)
+
+    wikisearch = ToolModule(
+        func=utils.webthink,
+        name="wikisearch",
+        description="Search[entity], which searches the exact entity on Wikipedia and returns the first paragraph if it exists. If not, it will return some similar entities to search."
+    )
+    wikilookup = ToolModule(
+        func=utils.webthink,
```

Comment:
Integrate these common tools into LLM Reasoners, i.e., People should be able to use these tools with `from reasoners.tools import WikiLookUp`.